Implement Dynamic CORS Configuration (Fixes #2)

HarshitShukla-dev opened the pull request:
This pull request addresses issue #2 by introducing dynamic CORS configuration to the backend. The changes allow the application to read allowed origins from an environment variable defined in the `.env` file, enhancing flexibility for different deployment environments (development, staging, production).

#### **Key Changes:**
- **Dynamic Configuration**: Allowed CORS origins are now read from the `.env` file as a comma-separated string.
- **Parsing and Validation**: The origins are parsed and validated as `Uri` types, with any invalid URLs logged as warnings.
- **Type Handling**: Implemented conversion of `Uri` to `HeaderValue` for compatibility with the `AllowOrigin::exact()` method in the CORS configuration.
- **Error Handling**: Added error handling for potential parsing issues, ensuring that the application does not crash on invalid inputs.

### **Reminder:**
Please ensure to update the `.env` file with the `ALLOWED_ORIGINS` variable in the following format:

```
ALLOWED_ORIGINS="http://example.com,http://another-example.com,http://localhost:3000"
```

HarshitShukla-dev committed:
feat: Dynamic CORS configuration from .env with improved error handling

- Added support for configuring allowed CORS origins dynamically from the .env file as a comma-separated list
- Parsed and validated origins as `Uri`, converting them to `HeaderValue` for CORS configuration
- Implemented error handling for invalid origins and failed conversions to `HeaderValue`
- Enhanced CORS flexibility for multiple environments (development, staging, production)
- Added logging for invalid or unparseable origins to assist in debugging

Sidharth-Singh10 reviewed (changes requested):
@HarshitShukla-dev , load the environment variable in `constants.rs`. and then use the constant in `main.rs`.

HarshitShukla-dev committed:
- Updated .env.example file to have the ALLOWED_ORIGINS variable.
- implemented the reading of cors origins from the .env file by loading the env file in constants.rs and then using the constants utility to read the value of ALLOWED_ORIGINS in main.rs

HarshitShukla-dev committed:
Improved formatting to pass the fmt tests

HarshitShukla-dev commented:
> @HarshitShukla-dev , load the environment variable in `constants.rs`. and then use the constant in `main.rs`.

I’m sorry for the oversight. I've successfully loaded the environment variable in constants.rs and integrated the constant into main.rs. The ALLOWED_ORIGINS variable is now dynamically configured based on the values specified in the .env file. Additionally, I've added error handling to ensure that the application gracefully handles any issues related to the environment variable.

If you have any further suggestions or need additional changes, please let me know!

Thanks!

HarshitShukla-dev committed:
Fixed the spelling of ALLOWED_ORIGINS

Sidharth-Singh10 commented on .env.example:
```diff
@@ -5,4 +5,4 @@ SMTP_HOST="smtp.example.com"
 PASS_RESET_LINK="http://localhost:5173/reset-password"
-ALOOWED_ORIGIN="http://localhost:3000,http://yourdomain.com"
\ No newline at end of file
+ALLOWED_ORIGIN="http://localhost:3000,http://yourdomain.com"
```
The spelling here is different from what you used in `constant.rs`. 
Change it to `ALLOWED_ORIGINS`

Sidharth-Singh10 commented on src/utils/constants.rs:
```diff
@@ -6,23 +6,53 @@ lazy_static! {
+    };
+
+    pub static ref ALLOWED_ORIGINS: Vec<String> = {
+        dotenv().ok();
```
calling `dotenv().ok()` again isn't necessary.

Sidharth-Singh10 commented:
If `ALLOWED_ORIGINS` is not set, do not default it to a empty list, exit the program gracefully with a warning.
Pointed out some minor faults please correct those.

HarshitShukla-dev committed:
- Updated .env.example file to have the correct spelling of ALLOWED_ORIGINS
- Updated constants.rs to exit the program if the ALLOWED_ORIGINS environment variable is not set

HarshitShukla-dev commented:
@Sidharth-Singh10, Fixed the issues.

Sidharth-Singh10 commented:
@HarshitShukla-dev , thanks for the work!!!

Sidharth-Singh10 commented on Cargo.toml:
```diff
@@ -20,7 +20,7 @@ chrono = "0.4.38"
 uuid = { version = "1.10.0", features = ["v4"] }
 bcrypt = "0.15.1"
-tower-http = { version = "0.5.2", features = ["cors"] }
+tower-http = { version = "0.6.1", features = ["cors"] }
```
please do not change versions of dependencies unless specified

HarshitShukla-dev commented:
Hey @Sidharth-Singh10, it'll great if you assign the hacktoberfest and gssoc labels to both the pr and the issue.
